### DISCUSSION TITLE
Example of overriding OnScopeSaving

### DISCUSSION THREAD

**Discussion by echamakiotis** [thepirat000/Audit.NET#529] [Q&A] [answered]
Hi,

I'm in the process of integrating Audit.EntityFramework in my project; I'm storing the audit output in the same DB as the real entities, using the EntityFramework data provider, and I need to ensure that both the real entity and the audit entity are saved within the same transaction. I understand that I need to override OnScopeSaving and use NullDataProvider instead of EntityFramework data provider, but I have no idea how to do the "custom log saving" mentioned in the documentation. @thepirat000, can you - or anyone else - please provide a working example?

Thank you

**Comment by thepirat000** [comment]
Hi. 

It really depends on your configuration and EF version.

I'm assuming that you:
- Are using the latest EF Core version.
- Want both saving operations (real entities & audit entities) to happen within the same transaction.
- Are using the same DB Context instance for both saving operations (i.e. you don't call `UseDbContext<T>(...)` on the config)
- **Don't want** to save an audit if the real entity saving fails,

Then my recommendation is to:

- Add two custom actions, one for `OnScopeCreated` to begin the transaction and another one on `OnEventSaved` to commit the transaction.
- Configure your DataProvider as usual with `Audit.Core.Configuration.Setup().UseEntityFramework(...)`

So, on your startup, this is how you can configure the custom actions to begin / commit the transaction:

```c#
Audit.Core.Configuration.Setup().UseEntityFramework(/*your custom configuration*/);

// Begin a transaction for each saving operation
Audit.Core.Configuration.AddCustomAction(ActionType.OnScopeCreated, async scope =>
{
    await scope.GetEntityFrameworkEvent().GetDbContext().Database.BeginTransactionAsync();
});

// Commit the transaction for each saving operation
Audit.Core.Configuration.AddCustomAction(ActionType.OnEventSaved, async scope =>
{
    await scope.GetEntityFrameworkEvent().GetDbContext().Database.CommitTransactionAsync();
});
```

In this way, both saving operations will be performed within the same transaction.

Note you cannot use the overrides on your `AuditDbContext`, since there is no virtual `OnEventSaved()` method to override. I will add that method on a future version, but the custom actions should have the exact same effect.

**Reply by echamakiotis** [reply]
Hi @thepirat000,

I need to ensure that either both the entity and the audit entry are saved, or none. So no audit entry should be saved if the real entity saving fails and vice-versa, the real entity is also not saved, if the audit entry could not be saved. Does the above approach provide this?

Thank you

**Reply by thepirat000** [reply]
Yes, since you use the same DbContext instance for both operations, that should cover your needs.

--

If real entity saving fails, then:

- The operation for the audit saving **will fail** since you use the same DbContext instance and it will be in a faulted state. That's why you don't need an explicit call to `RollbackTransaction` (the OnEventSaved action will not be called).   

If real entity saving succeeds, but audit saving fails, then:

- The transaction is never committed and will be rolled back when disposed (the OnEventSaved action will not be called)

If both saving operations succeed, then:

- The transaction is committed on the OnEventSaved action.

--

Nevertheless, you can change the OnEventSaved action to evaluate the Success property and rollback in such case, i.e.:

```c#
Audit.Core.Configuration.AddCustomAction(ActionType.OnEventSaved, async scope =>
{
    if (scope.GetEntityFrameworkEvent().Success)
    {
        await scope.GetEntityFrameworkEvent().GetDbContext().Database.CommitTransactionAsync();
    }
    else
    {
        await scope.GetEntityFrameworkEvent().GetDbContext().Database.RollbackTransactionAsync();
    }
});
```

**Comment by thepirat000** [comment, marked as answer]
Note that an OnScopeSaved() virtual method was added to the AuditDbContext
